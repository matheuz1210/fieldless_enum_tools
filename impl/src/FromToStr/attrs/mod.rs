use syn::{spanned::Spanned, Error, Lit, Meta, MetaList, NestedMeta};

use crate::utils;

macro_rules! try_get {
    ($attr:ident; $($field:ident => $typ:ty),*) => {
        if !$attr.path.is_ident("fromtostr") {
            continue;
        };
        let list: ::syn::MetaList = $attr.parse_args()?;

        $(if list.path.is_ident(stringify![$field]) {
            if $field.is_none() {
                $field = Some(<$typ>::from_attr(list)?);
                continue;
            };
            return Err(::syn::Error::new_spanned(list, "duplicate attribute!"));
        })*
        return Err(::syn::Error::new_spanned(list, "unknown attribute!"));
    };
}

macro_rules! malformed_err {
    ($span:ident, $($l:literal)*) => {
        move || {
            ::syn::Error::new(
                $span,
                concat!["malformed attribute, expected #[fromtostr(", $($l),*, ")]"],
            )
        }
    };
}

pub mod inner;
pub mod outer;

pub enum FormatCase {
    /// keep it as is
    None,
    /// "TwoWords" => "twowords"
    Lower,
    /// "TwoWords" => "twowords"
    Upper,
    /// "TwoWords" => "twoWords"
    Camel,
    /// "TwoWords" => "Two{separator}Words"
    Delimited { sep: String },
    /// "TwoWords" => "two{separator}words"
    DelimitedLower { sep: String },
    /// "TwoWords" => "TWO{separator}WORDS"
    DelimitedUpper { sep: String },
    /// "TwoWords" => "two_words"
    Snake,
    /// "TwoWords" => "TWO_WORDS"
    ScreamingSnake,
    /// "TwoWords" => "two-words"
    Kebab,
    /// "TwoWords" => "TWO-WORDS"
    ScreamingKebab,
    /// "TwoWords" => "Two-Words"
    Train,
    /// "TwoWords" => "Two_Words"
    PascalSnake,
    /// "TwoWords" => "two_Words"
    CamelSnake,
}

impl FormatCase {
    pub fn from_attr(attr: &MetaList, malformed_err: impl FnOnce() -> Error) -> syn::Result<Self> {
        let attr_span = attr.span();
        let mut separator = None;
        let mut style = None;

        for nested in &attr.nested {
            let nv = match nested {
                NestedMeta::Meta(Meta::NameValue(nv)) => nv,
                _ => return Err(malformed_err()),
            };

            match (
                utils::opt_as_deref(&nv.path.get_ident().map(ToString::to_string)),
                &nv.lit,
            ) {
                (Some("separator"), Lit::Str(sep)) => separator = Some(sep.value()),
                (Some("style"), Lit::Str(sty)) => style = Some((sty.value(), sty.span())),
                _ => return Err(malformed_err()),
            }
        }

        let (style, not_style_error) = match style {
            Some(a) => {
                let e = a.1;
                (a.0, move |s| {
                    Error::new(e, format!("{} not an avaliable style!", s))
                })
            }
            None => return Err(malformed_err()),
        };
        let no_sep_error =
            || Error::new(attr_span, r#"style = "delimited" needs a separator value"#);
        Ok(match style.trim() {
            "none" | "PascalCase" => FormatCase::None,
            "lower" => FormatCase::Lower,
            "UPPER" => FormatCase::Upper,
            "snake" => FormatCase::Snake,
            "SCREAMING_SNAKE" => FormatCase::ScreamingSnake,
            "kebab" => FormatCase::Kebab,
            "SCREAMING-KEBAB" => FormatCase::ScreamingKebab,
            "camel" => FormatCase::Camel,
            "camel_Snake" => FormatCase::CamelSnake,
            "Pascal_Snake" => FormatCase::PascalSnake,
            "Train" => FormatCase::Train,
            "delimited" => {
                if let Some(sep) = separator {
                    FormatCase::Delimited { sep }
                } else {
                    return Err(no_sep_error());
                }
            }
            "delimitedlower" => {
                if let Some(sep) = separator {
                    FormatCase::DelimitedLower { sep }
                } else {
                    return Err(no_sep_error());
                }
            }
            "DELIMITEDUPPER" => {
                if let Some(sep) = separator {
                    FormatCase::DelimitedUpper { sep }
                } else {
                    return Err(no_sep_error());
                }
            }
            s => return Err(not_style_error(s)),
        })
    }

    pub fn format(&self, s: &str) -> String {
        match self {
            FormatCase::None => s.to_owned(),
            FormatCase::Lower => s.to_lowercase(),
            FormatCase::Upper => s.to_uppercase(),
            FormatCase::Camel => s[..1].to_lowercase() + &s[1..],
            FormatCase::Delimited { sep } => FormatCase::delimit(sep, s),
            FormatCase::DelimitedLower { sep } => FormatCase::delimit(sep, s).to_lowercase(),
            FormatCase::DelimitedUpper { sep } => FormatCase::delimit(sep, s).to_uppercase(),
            FormatCase::Train => FormatCase::delimit("-", s),
            FormatCase::PascalSnake => FormatCase::delimit("_", s),
            FormatCase::CamelSnake => {
                let s = FormatCase::delimit("_", s);

                s[..1].to_lowercase() + &s[1..]
            }
            FormatCase::Snake => FormatCase::delimit("_", s).to_lowercase(),
            FormatCase::Kebab => FormatCase::delimit("-", s).to_lowercase(),
            FormatCase::ScreamingSnake => FormatCase::delimit("_", s).to_uppercase(),
            FormatCase::ScreamingKebab => FormatCase::delimit("-", s).to_uppercase(),
        }
    }

    pub fn delimit(sep: &str, s: &str) -> String {
        let mut string = String::with_capacity(s.len() + (4 * sep.len()));
        let mut chars = s.chars().peekable();
        // don't try to add an separator at the start
        string.push(chars.next().unwrap());

        let mut prev_is_upper = None;
        while let Some(ch) = chars.next() {
            let is_upper = ch.is_uppercase();
            let next_is_upper = chars.peek().map(|ch| ch.is_uppercase());

            // if this character is upper, and neither the previous and next is upper, add separator
            if is_upper && (next_is_upper == Some(false) || prev_is_upper == Some(false)) {
                string.push_str(sep);
            }
            string.push(ch);
            prev_is_upper = Some(is_upper);
        }

        string.shrink_to_fit();
        string
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn format() {
        use FormatCase::*;
        let orig = "TwoWords";

        assert_eq!(Lower.format(orig), "twowords");
        assert_eq!(Upper.format(orig), "TWOWORDS");
        assert_eq!(Camel.format(orig), "twoWords");
        assert_eq!(Snake.format(orig), "two_words");
        assert_eq!(ScreamingSnake.format(orig), "TWO_WORDS");
        assert_eq!(Kebab.format(orig), "two-words");
        assert_eq!(ScreamingKebab.format(orig), "TWO-WORDS");
        assert_eq!(Train.format(orig), "Two-Words");
        assert_eq!(PascalSnake.format(orig), "Two_Words");
        assert_eq!(CamelSnake.format(orig), "two_Words");

        // check if initialisms work
        assert_eq!(Train.format("HTTPRequest"), "HTTP-Request");
    }
}
