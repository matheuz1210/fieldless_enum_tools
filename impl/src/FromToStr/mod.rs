mod attrs;
use std::collections::HashMap;

use attrs::{inner::Attrs as InnerAttrs, outer::Attrs as OuterAttrs};
use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::Error;

pub fn main(input: syn::DeriveInput) -> syn::Result<TokenStream> {
    let data = crate::utils::check_if_fieldless_enum("FromToStr", input.data)?;
    let outer_attr = OuterAttrs::from_attrs(&input.attrs)?;

    let mut fmtd = Vec::with_capacity(data.variants.len());
    for var in data.variants {
        let inner_attr = InnerAttrs::from_attrs(&var.attrs)?;

        fmtd.push(FormattedVariant::new(var.ident, inner_attr, &outer_attr));
    }
    check_if_duplicate(&fmtd)?;

    let typ = input.ident;
    let mut tree = Impl::default_impl(&typ, fmtd.as_slice());

    for imp in &Impl::default() {
        let imp = *imp;
        if !outer_attr.should_skip(imp) {
            tree.extend(imp.quote_impl(&typ));
        }
    }

    Ok(tree)
}

fn check_if_duplicate(fmtd: &[FormattedVariant]) -> syn::Result<()> {
    let mut hashes = HashMap::new();

    for (str, span1) in fmtd.iter().flat_map(FormattedVariant::iter) {
        // if there was already this string on the hashmap, error
        if let Some(span2) = hashes.insert(str, span1) {
            let mut error = Error::new(*span1, format!("duplicate value! both are [{}]", str));
            error.combine(Error::new(
                *span2,
                format!("duplicate value! both are [{}]", str),
            ));

            return Err(error);
        }
    }
    Ok(())
}

pub struct FormattedVariant {
    original: Ident,
    formatted: (String, Span),
    aliases: Vec<(String, Span)>,
}

impl FormattedVariant {
    pub fn new(ident: Ident, inner_attr: InnerAttrs, outer_attr: &OuterAttrs) -> Self {
        Self {
            formatted: {
                let span = ident.span();
                let ident = ident.to_string();
                if let Some(ren) = inner_attr.rename {
                    use attrs::inner::Rename;
                    match ren {
                        Rename::Renamed(ren) => ren,
                        Rename::Format(f) => (f.format(&ident), span),
                    }
                } else if let Some(ref f) = outer_attr.format {
                    (f.format(&ident), span)
                } else {
                    (ident, span)
                }
            },
            original: ident,
            aliases: inner_attr.aliases.map(|a| a.0).unwrap_or_default(),
        }
    }
    pub fn iter(&self) -> impl Iterator<Item = &(String, Span)> {
        self.aliases.iter().chain(std::iter::once(&self.formatted))
    }
}

#[derive(PartialEq, Eq, Clone, Copy)]
pub enum Impl {
    AsRefStr,
    IntoString,
    Display,
    FromStr,
    TryFromString,
    Serialize,
    Deserialize,
}

impl Impl {
    pub const fn default() -> [Impl; 7] {
        [
            Impl::AsRefStr,
            Impl::IntoString,
            Impl::Display,
            Impl::FromStr,
            Impl::TryFromString,
            Impl::Serialize,
            Impl::Deserialize,
        ]
    }

    pub fn default_impl(typ: &Ident, formatted: &[FormattedVariant]) -> TokenStream {
        let mut as_str = TokenStream::new();
        let mut from_str = TokenStream::new();

        for f in formatted {
            let ident = &f.original;
            let formatted = &f.formatted.0;

            as_str.extend(quote! {
                #typ::#ident => #formatted,
            });

            let strings = f.iter().map(|s| &s.0);
            from_str.extend(quote! {
                #(#strings )|* => #typ::#ident,
            });
        }

        quote! {
            impl #typ {
                #[doc(hidden)]
                fn __as_str(&self) -> &'static str {
                    match self {
                        #as_str
                    }
                }
                #[doc(hidden)]
                fn __from_str(s: &str) -> Result<Self, ()> {
                    Ok(match s {
                        #from_str
                        _ => return Err(())
                    })
                }
            }
        }
    }

    pub fn quote_impl(self, typ: &Ident) -> TokenStream {
        match self {
            Impl::AsRefStr => quote! {
                impl ::core::convert::AsRef<str> for #typ {
                    #[inline]
                    fn as_ref(&self) -> &'static str {
                        self.__as_str()
                    }
                }
            },

            Impl::IntoString => {
                let string = quote! {::fieldless_enum_tools::__internal::String};
                quote! {
                    fieldless_enum_tools::if_alloc_enabled! {
                        impl ::core::convert::Into<#string> for #typ {
                            #[inline]
                            fn into(self) -> #string {
                                self.__as_str().to_owned()
                            }
                        }
                    }
                }
            }

            Impl::Display => quote! {
                impl ::core::fmt::Display for #typ {
                    #[inline]
                    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
                        f.write_str(self.__as_str())
                    }
                }
            },

            Impl::FromStr => quote! {
                impl ::core::str::FromStr for #typ {
                    type Err = ();

                    #[inline]
                    fn from_str(s: &str) -> Result<Self, Self::Err> {
                       Self::__from_str(s)
                    }
                }
            },

            Impl::TryFromString => {
                let string = quote! {::fieldless_enum_tools::__internal::String};
                quote! {
                    ::fieldless_enum_tools::if_alloc_enabled! {
                        impl ::core::convert::TryFrom<#string> for #typ {
                            type Error = ();

                            #[inline]
                            fn try_from(s: #string) -> Result<Self, Self::Error> {
                                Self::__from_str(&s)
                            }
                        }
                    }
                }
            }

            Impl::Serialize => {
                let serde = quote! {::fieldless_enum_tools::__internal::serde};
                quote! {
                    ::fieldless_enum_tools::if_serde_enabled! {
                        impl #serde::Serialize for #typ {
                            fn serialize<S: #serde::Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
                                self.__as_str().serialize(ser)
                            }
                        }
                    }
                }
            }

            Impl::Deserialize => {
                let visitor = quote::format_ident!("__{}Visitor", typ);
                let serde = quote! {::fieldless_enum_tools::__internal::serde};
                quote! {
                    ::fieldless_enum_tools::if_serde_enabled! {
                        impl<'de> #serde::Deserialize<'de> for #typ {
                            fn deserialize<D: #serde::Deserializer<'de>>(de: D) -> Result<Self, D::Error> {
                                struct #visitor;
                                impl<'de> #serde::de::Visitor<'de> for #visitor {
                                    type Value = #typ;

                                    fn expecting(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
                                        f.write_str(concat!["an ", stringify!(#typ)])
                                    }

                                    fn visit_str<E: #serde::de::Error>(self, v: &str) -> Result<Self::Value, E>{
                                        <#typ>::__from_str(v).map_err(|_| E::invalid_value(#serde::de::Unexpected::Str(v), &self))
                                    }
                                }

                                de.deserialize_str(#visitor)
                            }
                        }
                    }
                }
            }
        }
    }
}
