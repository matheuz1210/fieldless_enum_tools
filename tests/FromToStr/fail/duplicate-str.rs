use fieldless_enum_tools::FromToStr;

#[derive(FromToStr)]
enum Enum {
    VariantOne,
    #[fromtostr(rename("VariantOne"))]
    VariantTwo,
}

fn main() {}
