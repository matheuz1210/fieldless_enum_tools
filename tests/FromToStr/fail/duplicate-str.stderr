error: duplicate value! both are [VariantOne]
 --> tests/FromToStr/fail/duplicate-str.rs:6:24
  |
6 |     #[fromtostr(rename("VariantOne"))]
  |                        ^^^^^^^^^^^^

error: duplicate value! both are [VariantOne]
 --> tests/FromToStr/fail/duplicate-str.rs:5:5
  |
5 |     VariantOne,
  |     ^^^^^^^^^^
